ykcompile/src/lib.rs

```diff
@@ -503,6 +507,109 @@ impl<TT> TraceCompiler<TT> {
         Ok(())
     }
 
+    fn c_checked_binop(
+        &mut self,
+        dest: &Place,
+        binop: &BinOp,
+        op1: &Operand,
+        op2: &Operand,
+    ) -> Result<(), CompileError> {
```

Comment:
This doesn't seem to write a tuple anywhere. `CheckedBinop` returns a tuple of `(is_overflow, result_value)`.

Reply:
Since this just compiles parts of our trace, which is then run later, we don't need to return anything here. In other words: w can't know at this point whether we will overflow or not, until we actually run the compiled trace.

Reply:
I don't think that's what Bjorn is saying (but I might be wrong): I think he's saying that if we don't return the tuple, there's no way for us to check whether `is_overflow` is true or not.

Reply:
It'd be a `JC` (jump if carry flag set) instruction in the trace, wouldn't it?

Reply:
> It'd be a JC (jump if carry flag set) instruction in the trace, wouldn't it?

Jump to what? How does that code know if it's the overflow/not-overflow case? One way or the other we have to choose an encoding and, presumably, the current Rust codegen defaults to writing a tuple to heap (and, I would guess furthermore, probably expects LLVM to optimise that away).

So in a sense the answer to your question could be "yes" but we just need a consistent encoding and we have to make sure it corresponds with what other parts of the rustc codegen are expecting.

Reply:
Actually, I think I understand what you mean now. In MIR, `checked_add`, writes the result to a tuple:
```
$8 = checked_add($7, 2u8)
$1 = ($8).0 // 0 holds the result, 1 the overflow flag
```
We are currently only writing the result to a normal `Local`. We definitely need to proper projections first before we can do that, which we currently don't have.

Reply:
Should we wait for a projections PR before going ahead with this PR then?

Reply:
> Jump to what? 

To a `ud2` instruction or similar.

But the missing detail here is that the check is MIR (and thus SIR) level, so in fact @bjorn3  is right. This should return a tuple eventually (once we have support for types that are no u64).

I'm currently wrestling with `fm`, otherwise the types branch is close...

Reply:
> This should return a tuple eventually (once we have support for types that are no u64).

For avoidance of confusion: this should compile code that writes the result and overflow flag to a tuple (so the function this comment is annotated will continue to not return anything).

Reply:
After some offline discussion with @vext01, we've reached the following conclusion:

We can't trace a program that overflows, since the interpreter would just crash during tracing. This means that the trace assumes things never overflow, so if they do we'll need guards which exit the trace gracefully. The guards would have to check the tuple returned by `checked_add`, so we will eventually need to write the overflow flag to that tuple. However, since we don't have blackholing we currently don't have guards either. So for the time being we will have to just crash the trace instead, which we can do with a jump to an `ud2` instructions as explained above by @vext01.

This means, even if do projections first and write the overflow flag, we won't use it anyway for the time being. Though I can be convinced to do projections first if @ltratt deems it necessary before we proceed with this PR.

Reply:
Am I right in thinking that we'd have to rewrite all this after projections is done anyway?

Reply:
Not really. There's only one place here which uses projections: https://github.com/softdevteam/yk/pull/86/files#diff-162e1dfdca14767450c768ad342d54f2R143. But even with projections that line shouldn't look all that different.

What we can do however, once we have projections, is write the overflow flag to the `checked_add` return tuple, but we won't be using it until we have blackholing.

Reply:
OK, I'm happy to go with what you and Edd think.

Reply:
Fixed in d434138